File: src/main/java/com/allaboutjava/chap09/Tutorial_01_Function_With_Future_Return_Type.java
package com.allaboutjava.chap09;

import io.vavr.concurrent.Future;

import java.util.function.Function;

import static io.vavr.API.printf;
import static io.vavr.API.println;
import static java.util.concurrent.TimeUnit.SECONDS;

/**
 * Method with future as return type.
 */
public class Tutorial_01_Function_With_Future_Return_Type {

    public static void main(String[] args) {

        println("Step 1: Define a function which returns a Future");
        Function<String, Future<Integer>> donutStock = __ ->  Future.of(() -> {
            // assume some long running database operation
            println("checking donut stock");
            return 10;
        });

        println("\nStep 2: Call method which returns a Future");
        // ici l'appel est bloquant
        Integer vanillaDonutStock = donutStock.apply("vanilla donut").await(5, SECONDS).get();
        printf("Stock of vanilla donut = %s", vanillaDonutStock);
    }
}

File: src/main/scala/com/allaboutscala/chap09/Tutorial_01_Function_With_Future_Return_Type.scala
package com.allaboutscala.chap09

import scala.concurrent.Future
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Await
import scala.concurrent.duration._
import scala.language.postfixOps

/**
 * Method with future as return type.
 */
object Tutorial_01_Function_With_Future_Return_Type extends App {

  println("Step 1: Define a function which returns a Future")
  def donutStock(donut: String): Future[Int] = Future {
    // assume some long running database operation
    println("checking donut stock")
    10
  }

  println("\nStep 2: Call method which returns a Future")
  val vanillaDonutStock = Await.result(donutStock("vanilla donut"), 5 seconds)
  println(s"Stock of vanilla donut = $vanillaDonutStock")
}
